### PR TITLE
(BIDS-2238) manage multiple validator notifications

### PR DESCRIPTION
**Pull request by MauserBitfly**
### <samp>🤖 Generated by Copilot at 474d99a</samp>

This pull request improves the functionality and performance of the validator watchlist and event subscription feature. It adds new functions to the `db` package to handle batch operations and context-aware queries for validators and subscriptions. It also refactors the `handlers` package to use these new functions and simplify the logic of parsing and validating user inputs. Additionally, it introduces a new feature that allows users to input either validator indices or public keys when adding or removing validators from their watchlist or subscriptions.

### CODE REVIEW

**Comment by D13ce** [comment]
This PR replaces https://github.com/gobitfly/eth2-beaconchain-explorer/pull/2394.

**Review by Eisei24 — CHANGES_REQUESTED** [review]
Everything works but please have a look at my points.

File: handlers/common.go [hunk 1]
```diff
@@ -556,6 +556,30 @@ func GetValidatorIndexFrom(userInput string) (pubKey []byte, validatorIndex uint
 	return
 }
 
+// GetValidatorIndexFrom gets the validator index from users input
```
[the remark this answers is not in the document]
Comment (MauserBitfly) [reply handlers/common.go]: fixed

File: db/subscriptions.go [hunk 2]
```diff
@@ -80,6 +80,38 @@ func RemoveFromWatchlist(userId uint64, validator_publickey string, network stri
...
 
+func RemoveFromWatchlistBatch(userId uint64, validator_publickeys []string, network string) error {
+	keys := [][]byte{}
+	for _, keyString := range validator_publickeys {
+		key, err := hex.DecodeString(keyString)
+		if err != nil {
+			return err
+		}
+		keys = append(keys, key)
+	}
```
[the remark this answers is not in the document]
Comment (MauserBitfly) [reply db/subscriptions.go:91]: ok

File: handlers/common.go [hunk 3]
```diff
@@ -556,6 +556,30 @@ func GetValidatorIndexFrom(userInput string) (pubKey []byte, validatorIndex uint
 	return
```
[the remark this answers is not in the document]
Comment (MauserBitfly) [reply handlers/common.go]: deleted

File: db/frontend.go [hunk 4]
```diff
@@ -288,6 +288,45 @@ func AddSubscription(userID uint64, network string, eventName types.EventName, e
...
+
+	name := string(eventName)
+	if network != "" {
+		name = strings.ToLower(network) + ":" + string(eventName)
+	}
+
+	numArgs := 6
+	valueStrings := make([]string, 0, len(eventFilter))
+	valueArgs := make([]interface{}, 0, len(eventFilter)*numArgs)
+
```
[the remark this answers is not in the document]
Comment (MauserBitfly) [reply db/frontend.go]: yea was too unlikely for me so I ignored it ... but after you mentioning it now I felt bad so I added some batching ;-)

File: db/frontend.go [hunk 5]
```diff
@@ -299,6 +338,16 @@ func DeleteSubscription(userID uint64, network string, eventName types.EventName
 	return err
 }
 
+func DeleteSubscriptionBatch(userID uint64, network string, eventName types.EventName, eventFilter []string) error {
+	name := string(eventName)
+	if network != "" && !types.IsUserIndexed(eventName) {
+		name = strings.ToLower(network) + ":" + string(eventName)
+	}
+
+	_, err := FrontendWriterDB.Exec("DELETE FROM users_subscriptions WHERE user_id = $1 and event_name = $2 and event_filter = ANY($3)", userID, name, eventFilter)
```
[the remark this answers is not in the document]
Comment (MauserBitfly) [reply db/frontend.go]: ok

File: db/frontend.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (Eisei24) [reply db/frontend.go]: Thanks. Tested it for 15k validators and it works.

File: db/frontend.go [hunk 6]
```diff
@@ -299,6 +360,16 @@ func DeleteSubscription(userID uint64, network string, eventName types.EventName
 	return err
 }
 
+func DeleteSubscriptionBatch(userID uint64, network string, eventName types.EventName, eventFilter []string) error {
+	name := string(eventName)
+	if network != "" && !types.IsUserIndexed(eventName) {
+		name = strings.ToLower(network) + ":" + string(eventName)
+	}
+
+	_, err := FrontendWriterDB.Exec("DELETE FROM users_subscriptions WHERE user_id = $1 and event_name = $2 and event_filter = ANY($3)", userID, name, pq.Array(eventFilter))
```
Comment (recy21) [inline db/frontend.go:369]: Please use uppercase AND, like in the other statements.

File: db/subscriptions.go [hunk 7]
```diff
@@ -80,6 +80,38 @@ func RemoveFromWatchlist(userId uint64, validator_publickey string, network stri
...
+	_, err = tx.Exec("DELETE FROM users_subscriptions WHERE user_id = $1 AND event_filter = ANY($2) AND event_name LIKE ($3 || '%')", userId, pq.StringArray(validator_publickeys), network+":")
+	if err != nil {
+		return fmt.Errorf("error deleting subscriptions for validator: %v", err)
+	}
+
+	tag := network + ":" + string(types.ValidatorTagsWatchlist)
+
+	_, err = tx.Exec("DELETE FROM users_validators_tags WHERE user_id = $1 AND validator_publickey = ANY($2) AND tag = $3", userId, pq.ByteaArray(keys), tag)
+	if err != nil {
+		return fmt.Errorf("error deleting validator from watchlist: %v", err)
```
Comment (recy21) [inline db/subscriptions.go:107]: You have to use %w, otherwise the error isn't correctly nested.

File: db/subscriptions.go [hunk 8]
```diff
@@ -80,6 +80,38 @@ func RemoveFromWatchlist(userId uint64, validator_publickey string, network stri
...
+	}
+	tx, err := FrontendWriterDB.Begin()
+	if err != nil {
+		return fmt.Errorf("error starting db transactions: %v", err)
+	}
+	defer tx.Rollback()
+
+	_, err = tx.Exec("DELETE FROM users_subscriptions WHERE user_id = $1 AND event_filter = ANY($2) AND event_name LIKE ($3 || '%')", userId, pq.StringArray(validator_publickeys), network+":")
+	if err != nil {
+		return fmt.Errorf("error deleting subscriptions for validator: %v", err)
```
Comment (recy21) [inline db/subscriptions.go:100]: You have to use %w, otherwise the error isn't correctly nested.

File: db/subscriptions.go [hunk 9]
```diff
@@ -80,6 +80,38 @@ func RemoveFromWatchlist(userId uint64, validator_publickey string, network stri
...
+	for _, keyString := range validator_publickeys {
+		key, err := hex.DecodeString(keyString)
+		if err != nil {
+			return err
+		}
+		keys = append(keys, key)
+	}
+	tx, err := FrontendWriterDB.Begin()
+	if err != nil {
+		return fmt.Errorf("error starting db transactions: %v", err)
```
Comment (recy21) [inline db/subscriptions.go:94]: You have to use %w, otherwise the error isn't correctly nested.

**Review by recy21 — CHANGES_REQUESTED** [review]
Checked SQL (looks fine, except style on 1 location) & Error (use %w 3x)
But not the logic itself.

File: db/frontend.go (hunk 6 above)
Comment (MauserBitfly) [reply db/frontend.go]: changed it now to upper case - also in the sql above and blow.

File: db/subscriptions.go (hunk 9 above)
Comment (MauserBitfly) [reply db/subscriptions.go]: wow first time I learn that there is a %w after half a year ;-) 
fixed

File: db/subscriptions.go (hunk 7 above)
Comment (MauserBitfly) [reply db/subscriptions.go]: fixed

File: db/subscriptions.go (hunk 8 above)
Comment (MauserBitfly) [reply db/subscriptions.go]: fixed

**Review by recy21 — APPROVED** [review]

**Pull request merged**